Packages/ClientRuntime/Sources/Pagination/PaginatorSequence.swift

```diff
@@ -0,0 +1,49 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+public struct PaginatorSequence<Input: PaginateToken, Output: HttpResponseBinding>: AsyncSequence where Input.Token: Equatable {
+    public typealias Element = Output
+    let input: Input
+    let inputKey: KeyPath<Input, Input.Token?>?
+    let outputKey: KeyPath<Output, Input.Token?>
+    let paginationFunction: (Input) async throws -> Output
+    
+    public init(input: Input,
+                inputKey: KeyPath<Input, Input.Token?>?,
+                outputKey: KeyPath<Output, Input.Token?>,
+                paginationFunction: @escaping (Input) async throws -> Output) {
+        self.input = input
+        self.inputKey = inputKey
+        self.outputKey = outputKey
+        self.paginationFunction = paginationFunction
+    }
+    
+    public struct PaginationIterator: AsyncIteratorProtocol {
+        public typealias Element = Output
+        var input: Input
+        let sequence: PaginatorSequence
+        var token: Input.Token? = nil
+        var isFirstPage: Bool = true
+        
+        public mutating func next() async throws -> Output? {
+            while token != nil || isFirstPage {
+                if let token = token {
+                    self.input = input.usingPaginationToken(token)
```

Comment:
I'm haven't dug into this yet, but you need to make sure that you handle both the case where the pagination token is totally unset as well as the case where it's set to `""`

Reply:
ah yes I forgot !token.isEmpty here cuz I was running into some compiler issues. let me add it back. good call out. and hm on the second one, ill add a check for that as well

Reply:
is token always a string? I think I remember why I didnt do this cuz I coudlnt figure out if token was always a string to make this check.

Reply:
handled both cases I think now